### PR TITLE
[camera] availableCameras() implementation for CameraX re-write

### PR DESCRIPTION
**Pull request by gmackall**
*Replace this paragraph with a description of what this PR is changing or adding, and why. Consider including before/after screenshots.*
This implements the availableCameras() method from the camera plugins platform interface. 

*List which issues are fixed by this PR. You must list at least one issue.*
https://github.com/flutter/flutter/issues/112660

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/main/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart [hunk 1]
```diff
@@ -16,6 +19,47 @@ class AndroidCameraCameraX extends CameraPlatform {
...
+      // TODO(gmackall): replace this with call to CameraInfo.getLensFacing when changes containing that method are available
+      if ((await backCameraSelector.filter(<CameraInfo>[cameraInfo]))
+          .isNotEmpty) {
+        cameraLensDirection = CameraLensDirection.back;
+      } else if ((await frontCameraSelector.filter(<CameraInfo>[cameraInfo]))
+          .isNotEmpty) {
+        cameraLensDirection = CameraLensDirection.front;
+      } else {
+        //Skip this CameraInfo as its lens direction is unknown
+        continue;
```
Comment (camsim99) [inline packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:50]: What cameras would this case catch? I assume external cameras, but I was under the impression that those are not supported by CameraX right now. For what it's worth, this makes sense, but just wondering if this is case we'd actually hit.

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart [hunk 2]
```diff
@@ -16,6 +19,47 @@ class AndroidCameraCameraX extends CameraPlatform {
...
+      }
+
+      cameraSensorOrientation = await cameraInfo.getSensorRotationDegrees();
+      cameraName = 'Camera $cameraCount';
```
Comment (camsim99) [inline packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:54]: Is this how Android named cameras with the method used in the old plugin?https://developer.android.com/reference/android/hardware/camera2/CameraManager#getCameraIdList()?

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/CameraSelectorHostApiImpl.java [hunk 3]
```diff
@@ -52,12 +52,9 @@ public List<Long> filter(@NonNull Long identifier, @NonNull List<Long> cameraInf
     }
 
     List<CameraInfo> filteredCameraInfos = cameraSelector.filter(cameraInfosForFilter);
-    final CameraInfoFlutterApiImpl cameraInfoFlutterApiImpl =
-        new CameraInfoFlutterApiImpl(binaryMessenger, instanceManager);
     List<Long> filteredCameraInfosIds = new ArrayList<Long>();
 
     for (CameraInfo cameraInfo : filteredCameraInfos) {
-      cameraInfoFlutterApiImpl.create(cameraInfo, result -> {});
```
Comment (camsim99) [inline packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/CameraSelectorHostApiImpl.java:60]: Why did you remove the creation of the `CameraInfo`s? I'm not positive this is safe to do.

**Review by camsim99 — CHANGES_REQUESTED** [review]
Overall looks good! I requested some changes and asked some questions. You will also need to create a test file for this class and write some tests for this method. Should be similar to the tests in the [old plugin](https://github.com/flutter/plugins/blob/86eda6992727690b48952ba6951133273706dca2/packages/camera/camera_android/test/android_camera_test.dart#L478).

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart (hunk 1 above)
Comment (gmackall) [reply packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:50]: I believe it won't catch any case at the moment, but in addition to the upcoming changes to add support for external cameras, they are adding a `LENS_FACING_UNKNOWN` (see [here](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:camera/camera-core/src/main/java/androidx/camera/core/CameraSelector.java;l=45?q=file:androidx%2Fcamera%2Fcore%2FCameraSelector.java%20class:androidx.camera.core.CameraSelector&ss=androidx%2Fplatform%2Fframeworks%2Fsupport)). 

We will probably want to change this code anyway because the change to expose lens facing info in CameraInfo will happen around the same time, but I wanted to add this just so nothing slips through.

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/CameraSelectorHostApiImpl.java (hunk 3 above)
Comment (gmackall) [reply packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/CameraSelectorHostApiImpl.java:60]: This was because we are passing in CameraInfos that have already been added to the instance manager, and can re-use the ones that remain after filtering. Closing after offline discussion

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart [hunk 4]
```diff
@@ -16,6 +19,47 @@ class AndroidCameraCameraX extends CameraPlatform {
...
+        CameraSelector.getDefaultFrontCamera();
+
+    CameraLensDirection cameraLensDirection;
+    int cameraCount = 0;
+    int cameraSensorOrientation;
+    String cameraName;
+
+    for (final CameraInfo cameraInfo in cameraInfos) {
+      // Determine the lens direction by filtering the CameraInfo
+      // TODO(gmackall): replace this with call to CameraInfo.getLensFacing when changes containing that method are available
```
[the remark this answers is not in the document]
Comment (gmackall) [reply packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:41]: Added! https://github.com/flutter/flutter/issues/118730

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart [hunk 5]
```diff
@@ -13,9 +18,67 @@ class AndroidCameraCameraX extends CameraPlatform {
     CameraPlatform.instance = AndroidCameraCameraX();
   }
 
+  ProcessCameraProvider? processCameraProvider;
+  CameraSelector? backCameraSelector;
+  CameraSelector? frontCameraSelector;
```
Comment (camsim99) [inline packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:23]: I suggest either making these private and keeping the setters or keeping them public (and marking them `@visibleForTesting`) and setting them directly.

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart [hunk 6]
```diff
@@ -13,9 +18,67 @@ class AndroidCameraCameraX extends CameraPlatform {
...
   Future<List<CameraDescription>> availableCameras() async {
-    throw UnimplementedError('availableCameras() is not implemented.');
+    final List<CameraDescription> cameraDescriptions = <CameraDescription>[];
+
+    processCameraProvider ??= await ProcessCameraProvider.getInstance();
+    final List<CameraInfo> cameraInfos =
+        await processCameraProvider!.getAvailableCameraInfos();
+
+    backCameraSelector ??= CameraSelector.getDefaultBackCamera();
+    frontCameraSelector ??= CameraSelector.getDefaultFrontCamera();
```
Comment (camsim99) [inline packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:36]: I suggest initializing these inline above versus in this method.

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart [hunk 7]
```diff
@@ -13,9 +18,67 @@ class AndroidCameraCameraX extends CameraPlatform {
...
+  CameraSelector? frontCameraSelector;
+
+
   /// Returns list of all available cameras and their descriptions.
   @override
   Future<List<CameraDescription>> availableCameras() async {
-    throw UnimplementedError('availableCameras() is not implemented.');
+    final List<CameraDescription> cameraDescriptions = <CameraDescription>[];
+
+    processCameraProvider ??= await ProcessCameraProvider.getInstance();
```
Comment (camsim99) [inline packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:31]: If you keep the setter for this, I'd use it here for consistency. You can move the null check to that method, too.

File: packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart (hunk 5 above)
Comment (camsim99) [reply packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:23]: (Optional suggestion) I'd suggest keeping the `CameraSelector`s public and remove the setters, but keep the `processCameraProvider` as is because if you move the null check to the setter for `processCameraProvider` like I commented below, it could be useful in this class, as well.

**Review by camsim99 — COMMENTED** [review]
Left some comments on structure, but after those changes and formatting are fixed, let's land this!

File: packages/camera/camera_android_camerax/test/android_camera_camerax_test.dart [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+        MockProcessCameraProvider();
+    final MockCameraSelector mockBackCameraSelector = MockCameraSelector();
+    final MockCameraSelector mockFrontCameraSelector = MockCameraSelector();
+    final MockCameraInfo mockFrontCameraInfo = MockCameraInfo();
+    final MockCameraInfo mockBackCameraInfo = MockCameraInfo();
+    AndroidCameraCameraX.registerWith();
+
+    //Set class level ProcessCameraProvider and camera selectors to created mocks
+    final AndroidCameraCameraX androidCameraCamerax = AndroidCameraCameraX();
+    androidCameraCamerax.setDefaultBackCameraSelector(mockBackCameraSelector);
```
Comment (camsim99) [inline packages/camera/camera_android_camerax/test/android_camera_camerax_test.dart:47]: The changes look good but I think you need to update the test! This file will also need the license at the top:

```
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.
```

Comment (gmackall) [reply packages/camera/camera_android_camerax/test/android_camera_camerax_test.dart]: Sorry about that! Fixed the tests, and added the license.

I also moved the null check back to the method and out of the process camera provider because

1. In my mind (as named) a setter should set the variable even if it is non-null, and
2. If the null check is in the setter, then the tests will all need to mock the host platform reply to getting the instance of ProcessCameraProvider, because we follow that code path when we pass it as an argument to the setter (even though we don't set, because in the setter we do the null check and find the _processCameraProvider already set to the mock.

Open to moving it back if you prefer it that way though!

Comment (camsim99) [reply packages/camera/camera_android_camerax/test/android_camera_camerax_test.dart]: I'm not tied to my approach. This should be fine!

**Review by camsim99 — APPROVED** [review]
LGTM! Just need to fix the presubmit tests!
